### PR TITLE
Enable Clang -Wall, -Wextra, and -Weverything

### LINKED ISSUE #11: Fix all Clang warnings
In GitLab by @valentjn on Apr 25, 2016, 17:37

Most of them are really useful and point to real bugs in the code.

### PR DESCRIPTION
**Pull request by valentjn**
Clang flags `-Wall` and `-Wextra` used to be enabled (and the corresponding warnings were fixed), but were accidentally disabled in e80fab3. Restore the flags and fix the emerging warnings. Even many warnings from `-Weverything` are fixed.

The question is: Which additional warning flags do we want to enable? When compiled with `-Weverything`, we currently have the following number of warnings:
```
{
  '-Wc++98-compat': 138244,
  '-Wshadow': 21351,
  '-Wdocumentation': 17673,
  '-Wdeprecated': 11496,
  '-Wweak-vtables': 8716,
  '-Wsign-conversion': 8271,
  '-Wpadded': 5332,
  '-Wdocumentation-deprecated-sync': 4455,
  '-Wc++98-compat-pedantic': 4163,
  '-Wfloat-equal': 1719,
  '-Wdouble-promotion': 1643,
  '-Wold-style-cast': 782,
  '-Wglobal-constructors': 629,
  '-Wdocumentation-unknown-command': 538,
  '-Wdocumentation-pedantic': 265,
  '-Wmissing-prototypes': 154,
  '-Wextra-semi': 133,
  '-Wc++98-compat-local-type-template-args': 127,
  '-Wcovered-switch-default': 102,
  '-Wexit-time-destructors': 54,
  '-Wnon-virtual-dtor': 51,
  '-Wundef': 23,
  '-Wunused-exception-parameter': 22,
  '-Wmissing-noreturn': 21,
  '-Wunreachable-code-return': 18,
  '-Wundefined-reinterpret-cast': 16,
  '-Wunreachable-code-break': 16,
  '-Wdocumentation-html': 12,
  '-Wmissing-variable-declarations': 11,
  '-Wunreachable-code': 4,
  '-Wimplicit-fallthrough': 1,
  '-Wconditional-uninitialized': 1,
  '-Wfloat-conversion': 1
}
```
Warnings like `-Wc++98-compat` are useless, but we should consider enabling some other warnings, see below.

Fixes #11.

### CODE REVIEW

**Comment by valentjn** [comment]
I've added `-Weverything` and fixed about half of the emerging warning types. The other half has been disabled and is as follows (together with most common reason):

* 7942 `-Wc++98-compat`: `auto`, `override`, etc.
* 1411 `-Wshadow`: constructors whose arguments have the same name as the member variables which are initialized with them
* 1111 `-Wc++98-compat-pedantic`: Boost test macros ("empty macro arguments")
* 801 `-Wsign-conversion`: converting `int` to `size_t`
* 629 `-Wglobal-constructors`: global variables in Boost tests
* 562 `-Wunused-parameter`: everywhere
* 369 `-Wdocumentation`: numerous things that Doxygen tolerates but Clang doesn't
* 263 `-Wpadded`: padding of many classes
* 245 `-Wdeprecated`: implicit copy constructors or assignment operators for polymorphic classes
* 235 `-Wfloat-equal`: comparison of `double`s with `==` or `!=`
* 154 `-Wmissing-prototypes`: missing function prototypes in Boost tests or examples
* 147 `-Wweak-vtables`: pure-header polymorphic classes
* 78 `-Wc++98-compat-local-type-template-args`: lambda functions
* 52 `-Wexit-time-destructors`: static members and variables
* 21 `-Wmissing-noreturn`: functions that check for dependencies with `#ifdef` and only throw an exception if the dependency is not configured
* 17 `-Wundef`: `#ifdef` dependency checking
* 15 `-Wdocumentation-unknown-command`: `\page`, `\image`, etc.
* 5 `-Wdocumentation-deprecated-sync`: missing deprecation attribute despite being deprecated in the Doxygen comment

The numbers are not comparable to the ones in the OP: The ones here are the unique warnings, whereas the ones in the OP contain duplicates (same type on the same line in the same file, e.g., when included multiple times in different translation units).

By using `-Weverything`, we might benefit from useful future warnings that the Clang team will add.

Of the warnings that are still unfixed, I regard only `-Wdeprecated` as somewhat useful, since it checks for the reasonable rule of three. The other warnings can mostly be ignored.

**Comment by valentjn** [comment]
I've fixed most of the additional warnings on Clang 8. The following two warnings have been disabled:

* 518 `-Wdisabled-macro-expansion`: Boost test macros
* 470 `-Wshadow-field`: same as `-Wshadow` before, only specific to shadowed fields of classes

Clang 8 warnings are currently not checked automatically by Jenkins, since Clang 8 is not installed on our machines. The easiest way would to be to wait they have been upgraded to Bionic (Clang 8 is in bionic-updates).

File: base/tests/BasisEval.cpp [hunk 1]
```diff
@@ -6,6 +6,8 @@
 #define BOOST_TEST_DYN_LINK
 #include <boost/test/unit_test.hpp>
 
+#include <limits>
```
Comment (leiterrl) [inline base/tests/BasisEval.cpp]: This is not needed anymore is it?

File: datadriven/src/sgpp/datadriven/algorithm/DBMatDatabase.cpp [hunk 2]
```diff
@@ -118,7 +119,7 @@ void DBMatDatabase::putDataMatrix(sgpp::base::GeneralGridConfiguration& gridConf
 }
 
 bool DBMatDatabase::gridConfigurationMatches(json::DictNode *node,
-      sgpp::base::GeneralGridConfiguration& gridConfig, size_t entry_num) {
+      sgpp::base:: GeneralGridConfiguration& gridConfig, size_t entry_num) {
```
Comment (leiterrl) [inline datadriven/src/sgpp/datadriven/algorithm/DBMatDatabase.cpp]: extra space added by accident?

File: datadriven/src/sgpp/datadriven/application/learnersgdeonoffparallel/MPIMethods.cpp [hunk 3]
```diff
@@ -758,8 +758,8 @@ void MPIMethods::processIncomingMPICommands(PendingMPIRequest &pendingMPIRequest
       break;
     case WORKER_SHUTDOWN_SUCCESS:std::cout << "Worker has acknowledged shutdown" << std::endl;
       break;
-    case NULL_COMMAND:std::cout << "Error: Incoming command has undefined command id" << std::endl;
-      throw sgpp::base::algorithm_exception("MPI_Packet with NULL command received.");
+    case nullptr_COMMAND:std::cout << "Error: Incoming command has undefined command id" << std::endl;
```
Comment (leiterrl) [inline datadriven/src/sgpp/datadriven/application/learnersgdeonoffparallel/MPIMethods.cpp]: I guess the MPI enums were not meant to be renamed

File: datadriven/src/sgpp/datadriven/application/learnersgdeonoffparallel/MPIRequestPool.cpp [hunk 4]
```diff
@@ -26,7 +26,7 @@ size_t MPIRequestPool::createMPIRequestHandle() {
 }
 
 void MPIRequestPool::deleteMPIRequestHandle(size_t handleIndex) {
-//            mpiRequestStorage[handleIndex] = MPI_REQUEST_NULL;
+//            mpiRequestStorage[handleIndex] = MPI_REQUEST_nullptr;
```
Comment (leiterrl) [inline datadriven/src/sgpp/datadriven/application/learnersgdeonoffparallel/MPIRequestPool.cpp]: another renamed MPI enum

File: datadriven/src/sgpp/datadriven/application/learnersgdeonoffparallel/NetworkMessageData.hpp [hunk 5]
```diff
@@ -29,7 +29,7 @@ enum MPI_COMMAND_ID {
   /**
    * Used to identify packets where the command id has not been set
    */
-      NULL_COMMAND,
+      nullptr_COMMAND,
```
Comment (leiterrl) [inline datadriven/src/sgpp/datadriven/application/learnersgdeonoffparallel/NetworkMessageData.hpp]: another renamed MPI enum

File: datadriven/src/sgpp/datadriven/operation/hash/OperationMPI/MPIEnviroment.cpp [hunk 6]
```diff
@@ -125,7 +125,7 @@ void MPIEnviroment::slave_mainloop(void) {
         std::cout << "Created input_comm on " << rank << std::endl;
       } else {
         MPI_Comm_create(MPI_COMM_WORLD, tmp_group, &tmp_comm);
-        if (tmp_comm != MPI_COMM_NULL) {
+        if (tmp_comm != MPI_COMM_nullptr) {
```
Comment (leiterrl) [inline datadriven/src/sgpp/datadriven/operation/hash/OperationMPI/MPIEnviroment.cpp]:
```suggestion
        if (tmp_comm != MPI_COMM_NULL) {
```

File: datadriven/src/sgpp/datadriven/operation/hash/OperationMPI/MPIEnviroment.cpp [hunk 7]
```diff
@@ -232,7 +232,7 @@ void MPIEnviroment::init_communicator(base::OperationConfiguration conf) {
 
   std::cout << "Created comm on " << MPIEnviroment::get_node_rank() << "\n";
   MPI_Comm_create(MPI_COMM_WORLD, node_neighbors, &communicator);
-  if (communicator == MPI_COMM_NULL) {
+  if (communicator == MPI_COMM_nullptr) {
```
Comment (leiterrl) [inline datadriven/src/sgpp/datadriven/operation/hash/OperationMPI/MPIEnviroment.cpp]:
```suggestion
  if (communicator == MPI_COMM_NULL) {
```

File: base/tests/BasisEval.cpp (hunk 1 above)
Comment (valentjn) [reply base/tests/BasisEval.cpp]: No. Good eye.

File: datadriven/src/sgpp/datadriven/application/learnersgdeonoffparallel/MPIMethods.cpp (hunk 3 above)
Comment (valentjn) [reply datadriven/src/sgpp/datadriven/application/learnersgdeonoffparallel/MPIMethods.cpp]: That was one of two types of things I did via automation. Never trust a machine if you can do it yourself ;)

**Pull request merged**